---
title: "blastn taxonomic analysis"
author: "Kimberly Ledger"
date: "2023-06-30"
output: html_document
---

taxonomic id of blastn output. this code is useful for assigning higher classifications (genus, family, etc) for ASVs that have multispecies hits.  

**important** - i output a fasta that included ALL ASVs from the *combined* sequence filtering analysis of the two MiSeq runs with the GOApcod_2021 samples (it lives here: "/genetics/edna/workdir/GOApcod_2021/combined/trimmed/filtered/outputs/myasvs.fasta").  since this fasta is very big and it would take a very long time to run blastn taxonomy on the entire thing, i filtered the fasta by the ASVs that made it through the read decontamination code (ASV_decontamination_GOApcod_NEW.Rmd) and filtered myasvs.fasta and saved this new fasta file with fewer asvs as "/genetics/edna/workdir/GOApcod_2021/combined/trimmed/filtered/outputs/myasvs_decontaminated_20230630.fasta". 

```{r}
x <- read.csv("/home/kimberly.ledger/GOApcod_2021/decontamination_output_20230630.csv")

asvs <- unique(x$ASV)

library(Biostrings)

fasta <- readDNAStringSet("/genetics/edna/workdir/GOApcod_2021/combined/trimmed/filtered/outputs/myasvs.fasta")

filtered_fasta <- fasta[names(fasta) %in% asvs] 

writeXStringSet(filtered_fasta, "/genetics/edna/workdir/GOApcod_2021/combined/trimmed/filtered/outputs/myasvs_decontaminated_20230630.fasta")
```

I then... 

Ran blastn on "myasvs_decontaminated_20230630.fasta" output for example: 
[kimberly.ledger@akc0ss-vu-134 outputs]$ nohup blastn -db nt -query myasvs_decontaminated_20230630.fasta -perc_identity 96 -qcov_hsp_perc 100 -num_threads 10 -out blastnresults_out_20230630 -outfmt '6 qseqid sseqid pident length mismatch gapopen qstart qend sstart send evalue bitscore sscinames staxids'

may need to retype - and ' because copy-paste format throws things off

Run taxonkit on blastn output
cat blastnresults_out_20230630 | taxonkit lineage -c -i 14 > blastn_tax_out_20230630
taxonkit reformat blastn_tax_out_20230630 -i 16 > blastn_taxlineage_20230630.txt

Now use this script to generate ASV id's from 'blastn_taxlineage_20230630.txt'


```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

```{r load-libraries}
library(dplyr)
library(tidyr)
library(stringr)
```


## load output from taxonkit 
```{r load-output-from-taxonkit}
taxonomy <-read.delim("/genetics/edna/workdir/GOApcod_2021/combined/trimmed/filtered/outputs/blastn_taxlineage_20230630.txt", header = FALSE, na.strings=c(""," ","NA"))

head(taxonomy)
```

## clean up the header a bit
```{r}
# use the full taxonomy rather than the seq id to collapse identical entries
tax_df <- taxonomy %>%
  filter(V4 > 100) %>% # make sure all retained matches are >100 bp
  select(-V2, -V5, -V6, -V7, -V14, -V8, -V9, -V10, -V11, -V12, -V15, -V16) %>%  #remove unnecessary columns
  group_by(V1, V17) %>% # group by the sequence key and the full taxonomy to reduce duplicate entries
  unique() %>% # doing that reduced the number of entries from 146k to 17k
  rename(qseqid=V1, perc_id=V3, length=V4, sscinames=V13, taxonomy=V17) %>% #rename headers
  filter(!str_detect(sscinames, "environmental")) %>% # filter out any environmental samples
  filter(!str_detect(sscinames, "synthetic")) #%>% # filter out any synthetic "samples"
```

## calcualate the number of ASV's with taxonomic assignment
```{r how-many-total-ASVs-w-taxonomy}
tax_df %>% 
  ungroup() %>%
  select(qseqid) %>%
  unique() %>%
  nrow
```

## reformat table 
```{r}
# formatting the taxonomy variables
taxon_df <- tax_df %>%
  filter(str_detect(taxonomy, ";")) %>%
  separate(taxonomy, into=c("kingdom", "phylum", "class", "order", "family", "genus", "species"), sep = ";")
```


# Sort and clean-up the results based on % identity

## find species that include numbers or "sp."
```{r}
taxon_df %>%
  filter(str_detect(sscinames, " x "))
```

## if there are hybrid sequences, remove them if other ID's with same genus are present 

for example: 
```{r}
taxon_df <- taxon_df %>%
  filter(species != "Oncorhynchus mykiss x Salmo salar") %>%
  filter(species != "Salvelinus fontinalis x Salvelinus malma")
```

## rename known species with new names 

for example: 
```{r}
taxon_df$species[taxon_df$species == "Gadus ogac"] <- "Gadus macrocephalus"
```

## remove mammal ASVs

depending on the primer set, removing additional taxonomic groups may be necessary 
```{r}
taxon_df <- taxon_df %>%
  filter(class != "Mammalia")
```


## figure out what ASVs are unambiguously a single species

```{r}
single_spp_seqs <- taxon_df %>% 
  group_by(qseqid) %>%
  add_tally(name = "n_taxa") %>%
  filter(n_taxa == 1)
```

## remove the single-species seqs from the dataframe and then rank the hits by % identity for the remaining seqs

```{r}
seq_id_diff <- taxon_df %>%
  anti_join(., single_spp_seqs) %>%  ## comment out this line if there are no single species ASVs
  select(-length) %>%
  group_by(qseqid, species, genus, family, order, class, phylum, kingdom) %>%
    mutate(seq_percID = max(perc_id)) %>%
    group_by(qseqid, species, genus, family, order, class, phylum, kingdom, seq_percID) %>%
  summarise(max(seq_percID)) %>% # take just the top hit for each taxon (for each sequence)
  select(-`max(seq_percID)`) %>%
  ungroup() %>%
  group_by(qseqid) %>%
      mutate(id_rank = rank(desc(seq_percID), ties.method = "min")) %>% # rank the taxonomic hits per sequence by % id
       mutate(top_perc = max(seq_percID)) %>% # designate the highest % id for the best taxonomic hit in each sequence (in some, but not all cases, this is 100%)   
      mutate(diff = top_perc - seq_percID) %>% # calculate the difference between the % identity of the top hit and each subsequent taxonomic hit
        arrange(diff)

seq_id_diff %>%
  filter(diff > 0)
```

Now you have the single best entry for each species for each sequence ranked and with the difference between the first and second ranked entries calculated.

For sequences with multiple top hits, where the difference between ranked taxa = 0, you will end up defaulting to genus- or family-level ID. We'll do the same for any sequences where the difference betweeen the first and second ranked taxa is < 2%.

## Figure out which differences are **> 1%** and eliminate those first
```{r}
# filter out any taxa that are >1% less matching identity than the top taxonomic hit for a given sequence
to_remove_low_perc_hits <- seq_id_diff %>%
  ungroup() %>%
  group_by(qseqid) %>%
  filter(diff > 1)

keepers <- seq_id_diff %>%
  anti_join(to_remove_low_perc_hits)
```

## so now determine whether ASVs should be assigned to genus, family, order, etc. 
```{r}
singletons <- keepers %>%
  select(qseqid) %>%
  tally() %>%
  filter(n == 1)

# these are the seqs that now have only a single match
singleton_df <- singletons %>%
  left_join(keepers) %>%
  select(-n) %>%
  bind_rows(single_spp_seqs) %>% # combine the single spp data
  mutate(taxonomic_level = "species") %>%
  mutate(taxon = species)

## Genus-level matches
# remove the singletons from the bigger df 
single_genus <- keepers %>%
  anti_join(singleton_df)%>% # at best, these should be genus-level matches
  group_by(qseqid, genus) %>%
  tally() %>%
  ungroup() %>%
  group_by(qseqid) %>%
  tally() %>%
  filter(n == 1) %>% # seqs that match a single genus
  select(-n) %>%
  left_join(., keepers) %>%
  mutate(taxonomic_level = "genus") %>%
  mutate(taxon = genus)
  

## Family-level matches
single_family <- keepers %>%
  anti_join(singleton_df)%>%
  anti_join(single_genus) %>%
  group_by(qseqid, family) %>%
  tally() %>%
  ungroup() %>%
  group_by(qseqid) %>%
  tally() %>%
  filter(n == 1) %>% # seqs that match a single family
  select(-n) %>%
  left_join(., keepers) %>%
  mutate(taxonomic_level = "family") %>%
  mutate(taxon = family)


## Order-level matches
single_order <- keepers %>%
  anti_join(singleton_df)%>%
  anti_join(single_genus) %>%
  anti_join(single_family) %>%
  group_by(qseqid, order) %>%
  tally() %>%
  ungroup() %>%
  group_by(qseqid) %>%
  tally() %>%
  filter(n == 1) %>% # seqs that match a single order
  select(-n) %>%
  left_join(., keepers) %>%
  mutate(taxonomic_level = "order") %>%
  mutate(taxon = order)


## Class-level matches
single_class <- keepers %>%
  anti_join(singleton_df)%>%
  anti_join(single_genus) %>%
  anti_join(single_family) %>%
  anti_join(single_order) %>%
  group_by(qseqid, class) %>%
  tally() %>%
  ungroup() %>%
  group_by(qseqid) %>%
  tally() %>% 
  filter(n == 1) %>% # seqs that match a single class
  select(-n) %>%
  left_join(., keepers) %>%
  mutate(taxonomic_level = "class") %>%
  mutate(taxon = class)


## Phylum-level matches
single_phylum <- keepers %>%
  anti_join(singleton_df)%>%
  anti_join(single_genus) %>%
  anti_join(single_family) %>%
  anti_join(single_order) %>%
  anti_join(single_class) %>%
  group_by(qseqid, phylum) %>%
  tally() %>%
  ungroup() %>%
  group_by(qseqid) %>%
  tally() %>% 
  filter(n == 1) %>% # seqs that match a single phylum
  select(-n) %>%
  left_join(., keepers) %>%
  mutate(taxonomic_level = "phylum") %>%
  mutate(taxon = phylum)

# Kingdom level matches
single_kingdom <- keepers %>%
  anti_join(singleton_df)%>%
  anti_join(single_genus) %>%
  anti_join(single_family) %>%
  anti_join(single_order) %>%
  anti_join(single_class) %>%
  group_by(qseqid, kingdom) %>%
  tally() %>%
  ungroup() %>%
  group_by(qseqid) %>%
  tally() %>% 
  filter(n == 1) %>% # seqs that match a single kingdom
  select(-n) %>%
  left_join(., keepers) %>%
  mutate(taxonomic_level = "kingdom") %>%
  mutate(taxon = kingdom)
```

## modify the singleton_df to include the right variable headers
```{r}
single_spp <- singleton_df %>%
  select(-perc_id, -length, -sscinames, -n_taxa) %>%
  mutate(taxonomic_level = "species") %>%
  mutate(taxon = species)
```

## recombine the full data set now that the appropriate level of taxonomy has been determined

```{r}
sorted_tax_df <- bind_rows(single_kingdom, single_phylum, single_class, single_order, single_family, single_genus, single_spp)
```

## Create output taxonomy data frames
```{r}
uncollapsed_taxonomy <- sorted_tax_df %>%
  select(-top_perc, -id_rank) %>%
  unique() %>%
  arrange(qseqid)

# and then collapse that down to just a single taxon per ASV
collapsed_taxonomy <- uncollapsed_taxonomy %>%
  select(qseqid, taxon, taxonomic_level) %>%
  unique() %>%
  rename(ASV = qseqid)
```


```{r}
write.csv(uncollapsed_taxonomy, "/genetics/edna/workdir/GOApcod_2021/combined/trimmed/filtered/outputs/asv_taxonomy_blastn_20230630.csv")
```
