---
title: "asv analysis of GOA pcod 2021 samples using MiFish"
author: "Kimberly Ledger"
date: "2023-05-29"
output: github_document
---

analysis of mifish sequences from May 9 2023 and May 12 2023 sequencing runs    
samples are from the GOA pcod survey in 2021  

this script uses decontaminated read counts. 

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

load libraries
```{r, warning=FALSE}
library(tidyverse)
library(ggplot2)
```

read in sample metadata
```{r}
metadata <- read.csv("/genetics/edna/workdir/GOApcod_2021/GOA2021_metadata_20230515.csv")

#illumina output changed "_" to "-"
metadata$Sample_ID <- gsub("_", "-", metadata$Sample_ID) 
```

read in taxonomic identification
- this is a table i created that integrated INSECT and blastn classifications
- i removed a few mammals and no rank ASVs --- GO BACK AND KEEP MARINE MAMMALS? 
```{r}
taxon <- read.csv("/genetics/edna/workdir/GOApcod_2021/combined/trimmed/filtered/outputs/asv_taxonomy_COMBINED.csv") %>%
  rename(ASV = representative)
```


read in decontaminated sample table
```{r}
asv_table <- read.csv("/home/kimberly.ledger/GOApcod_2021/decontaminated_reads.csv", row.names = 1)

asv_summary <- asv_table %>%
  group_by(ASV) %>%
  summarize(reads = sum(reads))
```

join taxon info to sample table 
```{r}
asv_w_id <- asv_summary %>%
  left_join(taxon, by = "ASV") %>%
  filter(!is.na(taxon))
```

plot the proportion of reads for a taxon assigning to individual ASV's 
```{r echo=FALSE}
asv_w_id %>%
  group_by(taxon) %>%
  filter(taxon != "NA") %>%
  mutate(sum=sum(reads)) %>%
  mutate(prop = reads/sum) %>%
  ggplot(aes(x=taxon, y=prop, fill=ASV)) +
  geom_bar(stat = "identity") + 
    theme_bw() +
  labs(
    y = "proportion of sequencing reads",
    x = "taxon",
    title = "ASV proportions") + 
  theme(
    axis.text.x = element_text(angle = 90, hjust = 0.95),
    legend.text = element_text(size = 8),
    legend.key.size = unit(0.3, "cm"),
    legend.position = "none",
    legend.title = element_blank()
  )
```

okay, so some taxa (especially species-level) are only made up of one ASV, while other taxa have several ASVs


asv summary table - only keeping order and below because everything above that (i.e. class) is identical 
```{r}
asv_id_table <- asv_w_id %>%
  select(ASV, order, family, genus, species, taxon, reads) %>%
  group_by(taxon) %>%
  filter(taxon != "NA") %>%
  mutate(sum=sum(reads)) %>%
  mutate(prop = reads/sum) %>%
  arrange(order, family, genus, species)
```

arrange taxons in a way that make sense instead of alphabetical 
```{r}
asv_order <- asv_w_id %>%
  arrange(order, family, genus, species) %>%
  select(taxon) %>%
  unique
```

now, join taxon and asv tables (with decontaminated read counts)
```{r}
asv_id <- asv_w_id %>%
  select(ASV, taxon)

read_summary <- asv_table %>%
  left_join(asv_id, by = "ASV") %>%
  filter(reads > 0) %>%
  filter(taxon != "NA") %>%
  group_by(Sample_ID, taxon) %>%
  summarise(total_read_count = sum(reads)) %>%
  pivot_wider(names_from = "taxon", values_from = "total_read_count") %>%
  replace(is.na(.), 0)
```

join to metadata  - UPDATE COLUMN NUMBERS BELOW depending on input read_summary data
```{r}
join <- metadata %>%
  left_join(read_summary, by = c("Sample_ID"))

join_long <- join %>%
  pivot_longer(cols = 20:62, names_to = "taxon", values_to = "reads") %>%
  mutate(reads = ifelse(is.na(reads), 0, reads))

join_long$Sample_ID <- as.factor(join_long$Sample_ID)
join_long$pcr_replicate <- as.factor(join_long$pcr_replicate)
join_long$extraction_ID <- as.factor(join_long$extraction_ID)
join_long$extraction_replicate <- as.factor(join_long$extraction_replicate) 
join_long$run <- as.factor(join_long$run)
join_long$sample_type <- as.factor(join_long$sample_type)
join_long$taxon <- as.factor(join_long$taxon)

summary(join_long)
```

relevel the taxon factor
```{r}
desired_order <- asv_order$taxon

join_long <- join_long %>%
  mutate(taxon = factor(taxon, levels = desired_order))

levels(join_long$taxon)
```

make a table with taxons and total read counts 
```{r}
join_long %>%
  group_by(taxon) %>%
  summarise(total_reads = sum(reads)) %>%
  arrange(desc(total_reads))
```



# now let's check out data! 

since reads were decontaminated it doesn't make sense to plot positive controls, pcr blanks, or extraction blanks... 
but remember there were lots of O. nerka reads in e00562-A that i did not incorporate into the decontamination process. 

## any sequences in field blanks? 
```{r echo=FALSE}
join_long %>% 
  filter(sample_type == "field_blank") %>%
  ggplot(aes(x=Sample_ID, y=reads, fill=taxon)) +
  geom_bar(stat = "identity") + 
  facet_wrap(~extraction_ID, scales = 'free', ncol = 3) +
  theme_bw() +
  labs(
    y = "sequencing reads",
    x = "sample",
    title = "field blanks") + 
  theme(
    axis.text.x = element_text(angle = 90, hjust = 0.95),
    legend.text = element_text(size = 8),
    legend.key.size = unit(0.3, "cm"),
    legend.position = "none",
    legend.title = element_blank()
  )
```

well, unfortunately only three field blanks were taken during this entire survey.  

what are the read counts in the field blanks?
```{r}
field_nc_reads <- join_long %>%
  filter(sample_type == "field_blank") %>% 
  filter(reads > 0) %>%
  select(Sample_ID, taxon, reads) %>%
  pivot_wider(names_from = "Sample_ID", values_from = "reads") 

field_nc_reads
```

some evidence for field contamination. 

well, well. will need to think about how to account/deal with this. 


create some read summaries 
```{r}
join_long %>%
  group_by(sample_type, run, Sample_ID) %>%
  summarise(total_reads = sum(reads)) %>%
  group_by(sample_type, run) %>%
  summarise(min_reads = min(total_reads),
            quant1_reads = quantile(total_reads, 0.25),
            median_reads = median(total_reads),
            mean_reads = mean(total_reads),
            quant3_reads = quantile(total_reads, 0.75),
            max_reads = max(total_reads))
```

not a huge difference in read counts for pc and samples between run A and B so I won't worry about any batch effects for now. 


take a quick first look at the library sizes (i.e. the number of reads) in each sample, as a function of whether that sample was a true positive sample or a negative control
```{r}
temp <- join_long %>%
  group_by(Sample_ID) %>%
  mutate(total_reads = sum(reads)) %>%
  arrange(total_reads) %>%
  select(Sample_ID, extraction_ID, sample_type, dna_conc, total_reads) %>%
  unique()

temp$Index <- seq(nrow(temp))

ggplot(temp, aes(x=Index, y = total_reads, color = sample_type)) + 
         geom_point(alpha = 0.5)
```

plot total reads by DNA concentration for samples 
```{r}
temp %>%
  filter(sample_type == "sample") %>%
  ggplot(aes(x=dna_conc, y = total_reads)) + 
         geom_point(alpha = 0.5)
```

hmm.. perhaps i should flag samples that have no reads post decontamination but have a DNA concentration of 2ng/ul or greater?? 

```{r}
temp %>%
  filter(dna_conc > 2) %>%
  filter(total_reads < 1) %>%
  arrange(Sample_ID)
```

maybe try to rerun these samples on the MiSeq??? 


alright, let me plot the legend for reference 
```{r echo=FALSE}
for_legend <- join_long %>%
  ggplot(aes(x=Sample_ID, y=reads, fill=taxon)) +
  geom_bar(stat = "identity") + 
  theme_bw() +
  labs(
    y = "sequencing reads",
    x = "sample",
    title = "PCR blanks") + 
  theme(
    axis.text.x = element_text(angle = 90, hjust = 0.95),
    legend.text = element_text(size = 7),
    legend.key.size = unit(0.2, "cm"),
    legend.position = "right",
    legend.title = element_blank()
  )

library(grid)
library(gridExtra) 

# Using the cowplot package
legend <- cowplot::get_legend(for_legend)

grid.newpage()
grid.draw(legend)
```

okay, now i will summarize samples by location (remember part of the decontamination steps included removing outlier site replicates)

```{r}
location_summary <- join_long %>%
  filter(sample_type == "sample") %>%
  group_by(Sample_ID) %>%
  mutate(sum=sum(reads)) %>%
  mutate(prop = reads/sum) %>%
  group_by(location1, taxon) %>%
  summarize(mean_prop = mean(prop))

location_summary$location1 <- as.factor(location_summary$location1)
```

```{r}
location_summary %>%
  ggplot(aes(x=location1, y=mean_prop, fill=taxon)) +
  geom_bar(stat = "identity") + 
  theme_bw() +
  labs(
    y = "proportion of sequencing reads",
    x = "location",
    title = "proportion of reads") + 
  theme(
    axis.text.x=element_text(angle = 90, hjust = 0.95), #remove x axis labels
    legend.text = element_text(size = 8),
    legend.key.size = unit(0.3, "cm"),
    legend.position = "none",
    legend.title = element_blank(),
  )
```

note the locations without columns is not missing data, there were not samples taken from locations of those values and my first couple attempts to remove them from the plot failed 






## now let's take a look at reads from the samples 
```{r echo=FALSE}
# join_long %>% 
#   filter(sample_type == "sample") %>%
#   ggplot(aes(x=Sample_ID, y=reads, fill=taxon)) +
#   geom_bar(stat = "identity") + 
#   scale_y_sqrt() +
#   theme_bw() +
#   labs(
#     y = "sequencing reads",
#     x = "sample",
#     title = "number of reads") + 
#   theme(
#     axis.text.x=element_blank(), #remove x axis labels
#     legend.text = element_text(size = 8),
#     legend.key.size = unit(0.3, "cm"),
#     legend.position = "none",
#     legend.title = element_blank(),
#   )
```

## and proportion of reads from the samples 

```{r}
# join_long %>% 
#   filter(sample_type == "sample") %>%
#   group_by(Sample_ID) %>%
#   mutate(sum=sum(reads)) %>%
#   mutate(prop = reads/sum) %>%
#   ggplot(aes(x=Sample_ID, y=prop, fill=taxon)) +
#   geom_bar(stat = "identity") + 
#   theme_bw() +
#   labs(
#     y = "proportion of sequencing reads",
#     x = "sample",
#     title = "proportion of reads") + 
#   theme(
#     axis.text.x=element_blank(), #remove x axis labels
#     legend.text = element_text(size = 8),
#     legend.key.size = unit(0.3, "cm"),
#     legend.position = "none",
#     legend.title = element_blank(),
#   )
```